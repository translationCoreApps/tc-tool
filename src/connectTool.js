import React from 'react';
import PropTypes from 'prop-types';
import {createProvider} from 'react-redux';
import {configureStore} from './state/store';
import {loadLocalization, setActiveLocale} from './state/actions/locale';
import BrokenScreen from './BrokenScreen';
import {
  getActiveLanguage,
  getLocaleLoaded,
  getTranslate,
} from './state/reducers';
import fs from 'fs-extra';

/**
 * This HOC initializes a store and locale for the tool.
 * It also specifies some required properties common to all tools.
 *
 * @param {string} [localeDir] - directory containing the interface locale files
 * @param {*} [reducer] - a custom reducer for the tool.
 * @return {function(*)}
 */
const connectTool = (localeDir, reducer) => {
  if (localeDir && typeof localeDir !== 'string') {
    throw Error(
      `Invalid parameter. Expected localeDir to be a string but found ${typeof localeDir} instead`);
  }

  return (WrappedComponent) => {
    const hasLocale = localeDir && fs.existsSync(localeDir);
    if(!localeDir) {
      console.warn('You should consider localizing this tool.');
    } else if (!hasLocale) {
      console.warn(`No locale found at ${localeDir}`);
    }

    const store = configureStore(reducer);
    // TRICKY: this will overwrite the default store context key
    // thus removing direct access to tC core's store which also uses the default key.
    const Provider = createProvider();

    /**
     * This container sets up the tool environment.
     *
     * @property {string} appLanguage - the app interface language code
     */
    class Tool extends React.Component {
      constructor(props) {
        super(props);
        this._isLoaded = this._isLoaded.bind(this);
        this.state = {
          broken: false,
          error: null,
          info: null,
        };
      }

      componentWillMount() {
        const {appLanguage} = this.props;
        if (hasLocale) {
          store.dispatch(loadLocalization(localeDir, appLanguage));
        }
        this.unsubscribe = store.subscribe(this.handleChange.bind(this));
      }

      componentWillUnmount() {
        this.unsubscribe();
      }

      handleChange() {
        this.forceUpdate();
      }

      componentDidCatch(error, info) {
        this.setState({
          broken: true,
          error,
          info,
        });
      }

      componentWillReceiveProps(nextProps) {
        // stay in sync with the application language
        if (hasLocale && nextProps.appLanguage !== this.props.appLanguage) {
          store.dispatch(setActiveLocale(nextProps.appLanguage));
        }
      }

      /**
       * Checks if the locale has finished loading
       * @return {bool}
       * @private
       */
      _isLoaded() {
        // TRICKY: if locale doesn't exist skip to finished loading
        return !hasLocale || getLocaleLoaded(store.getState());
      }

      /**
       * Returns the broken screen if one should be displayed
       * @return {*}
       * @private
       */
      _getBrokenScreen() {
        const {broken, error, info} = this.state;
        if (broken) {
          // TODO: log the error to the core app state so it will be included in feedback logs.
          // it would be best to pass a callback into this component for this purpose.
          // Perhaps we could also include the state of this tool.

          // TRICKY: we don't want to rely on tools localizing the title of the error page.
          // so we'll assume most people understand that "error" and some cryptic stack trace
          // means things broke.
          return <BrokenScreen title="ERROR"
                               error={error}
                               info={info}/>;
        } else {
          return null;
        }
      }

      render() {
        if (!this._isLoaded()) {
          // TODO: we could display a loading screen while the tool loads
          return null;
        }

        const brokenScreen = this._getBrokenScreen();
        if (brokenScreen) {
          return brokenScreen;
        } else {
          // TRICKY: load locale tools if enabled
          let translate = undefined;
          let currentLanguage = undefined;
          if (hasLocale) {
            translate = getTranslate(store.getState());
            const lang = getActiveLanguage(store.getState());
            currentLanguage = lang ? lang.code : undefined;
          }
          return (
            <Provider store={store}>
              <WrappedComponent
                translate={translate}
                currentLanguage={currentLanguage}
                {...this.props}
              />
            </Provider>
          );
        }
      }
    }

    /**
     * This defines the interface between tools and tC core.
     */
    Tool.propTypes = {
      currentToolViews: PropTypes.object.isRequired,
      resourcesReducer: PropTypes.object.isRequired,
      contextIdReducer: PropTypes.object.isRequired,
      appLanguage: PropTypes.string.isRequired,
    };
    return Tool;
  };
};

export default connectTool;
