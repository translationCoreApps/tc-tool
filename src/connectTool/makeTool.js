import React from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import { getLocaleLoaded } from '../state/reducers';
import { loadLocalization, setActiveLocale } from '../state/actions/locale';
import { makeToolProps } from './makeProps';

/**
 * Generates the tool component
 * @param WrappedComponent
 * @param {string} namespace - the tool namespace/id
 * @param store - the redux store
 * @param {string} localeDir - the directory where locale files will be loaded
 * @param {ToolApi} api - the tool's api
 * @return {Tool}
 */
export const makeTool = (
  WrappedComponent, namespace, store, localeDir = undefined,
  api = undefined) => {
  const hasLocale = Boolean(localeDir);
  const hasApi = Boolean(api);

  // checks if the locale has finished loading
  const isLocaleLoaded = () => {
    return !hasLocale || getLocaleLoaded(store.getState());
  };

  class Tool extends React.Component {
    constructor(props) {
      super(props);
      this.handleChange = this.handleChange.bind(this);
      this.toolDidUpdate = this.toolDidUpdate.bind(this);
    }

    componentWillMount() {
      const {appLanguage} = this.props;
      // TRICKY: if an api exists the locale will be loaded there.
      if (hasLocale && !hasApi) {
        store.dispatch(loadLocalization(localeDir, appLanguage));
      }
      this.unsubscribe = store.subscribe(this.handleChange);
      if (hasApi) {
        // Subscribe to the bare API's update trigger
        this.unsubscribeApi = api.subscribe(this.toolDidUpdate);
      }
    }

    componentWillUnmount() {
      this.unsubscribe();
      if (this.unsubscribeApi) {
        this.unsubscribeApi();
      }
    }

    /**
     * This is executed by the Tool Api in order to trigger re-rendering.
     * You could also override this to provide your own logic.
     */
    toolDidUpdate() {
      this.forceUpdate();
    }

    handleChange() {
      this.forceUpdate();
    }

    componentWillReceiveProps(nextProps) {
      // TODO: this is an anti-pattern. see https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html
      // stay in sync with the application language
      console.log(`Tool.componentWillReceiveProps(${namespace}) - hasLocale: ${hasLocale}, nextProps.appLanguage: ${nextProps.appLanguage}, this.props.appLanguage: ${this.props.appLanguage},`);
      if (hasLocale && nextProps.appLanguage !== this.props.appLanguage) {
        console.log(`Tool.componentWillReceiveProps(${namespace}) - updating locale`);
        store.dispatch(setActiveLocale(nextProps.appLanguage));
      }
    }

    render() {
      if (!isLocaleLoaded()) {
        // TODO: we could display a loading screen while the tool loads
        return null;
      } else {
        const toolProps = makeToolProps(this.props, namespace, store.getState(),
          hasLocale);
        const componentProps = {
          ...this.props, // TODO: this is deprecated
          toolApi: api, // TODO: this is deprecated
          ...toolProps,
          tool: {
            ...toolProps.tool,
            api: api
          }
        };

        return (
          <Provider store={store}>
            <WrappedComponent
              {...componentProps}
            />
          </Provider>
        );
      }
    }
  }

  Tool.propTypes = {
    appLanguage: PropTypes.string.isRequired,
    writeProjectData: PropTypes.func.isRequired,
    readProjectData: PropTypes.func.isRequired,
    readProjectDataSync: PropTypes.func.isRequired,
    deleteProjectFile: PropTypes.func.isRequired,
    projectDataPathExists: PropTypes.func.isRequired,
    projectDataPathExistsSync: PropTypes.func.isRequired
  };

  return Tool;
};
