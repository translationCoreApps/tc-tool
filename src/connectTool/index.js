import {configureStore} from '../state/store';
import {getActiveLanguage, getTranslate} from '../state/reducers';
import fs from 'fs-extra';
import ApiController from '../api/ApiController';
import {makeTool} from './makeTool';

/**
 * Builds the locale tools
 * @param state
 * @return {*}
 */
export const makeLocaleProps = (state) => {
  const translate = getTranslate(state);
  const lang = getActiveLanguage(state);
  const currentLanguage = lang ? lang.code : undefined;
  return {
    translate,
    currentLanguage
  };
};

/**
 * This HOC initializes a store and locale for the tool.
 * It also specifies some required properties common to all tools.
 *
 * @param namespace
 * @param options
 * @return {function(*)}
 */
const connectTool = (namespace, options = {}) => {

  /**
   * @param {string} [localeDir] - directory containing the interface locale files
   * @param {*} [reducer] - a custom reducer for the tool.
   * @param {[]} [middlewares] - an array of middleware to inject into the store.
   * @param {*} [api] - The tool's api class
   */
  const {localeDir = undefined, reducer = undefined, middlewares = undefined, api = undefined} = options;

  if (localeDir && typeof localeDir !== 'string') {
    throw Error(
      `Invalid parameter. Expected localeDir to be a string but found ${typeof localeDir} instead`);
  }

  return (WrappedComponent) => {
    console.info('Booting tool...');
    const hasLocale = localeDir && fs.existsSync(localeDir);
    if (!localeDir) {
      console.warn('You should consider localizing this tool.');
    } else if (!hasLocale) {
      console.warn(`No locale found at ${localeDir}`);
    }

    const store = configureStore(reducer, middlewares);

    // wrap api in controller
    let controlledApi = undefined;
    if (api) {
      api.toString = () => namespace;
      controlledApi = new ApiController(api, store,
        hasLocale ? localeDir : undefined);
    }

    return {
      name: namespace,
      api: controlledApi,
      container: makeTool(WrappedComponent, store,
        hasLocale ? localeDir : undefined, api)
    };
  };
};

export default connectTool;
