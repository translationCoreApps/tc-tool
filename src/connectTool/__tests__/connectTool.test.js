jest.unmock('react-localize-redux');
jest.mock('../../state/actions/locale');
import React from 'react';
import connectTool from '../../connectTool';
import {configure, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import path from 'path';
import renderer from 'react-test-renderer';


class TestComponent extends React.Component {
  render() {
    return (
      <div>
        Hello World!
      </div>
    );
  }
}

class BrokenComponent extends React.Component {
  render() {
    throw Error('Something broke!');
  }
}

const localeDir = path.join(__dirname, './locale');

describe('props', () => {
  beforeAll(() => {
    configure({adapter: new Adapter()});
  });

  test('no params', () => {
    expect(connectTool.bind(null, 'my tool')).not.toThrow();
  });

  test('invalid localeDir param', () => {
    expect(connectTool.bind(null, 'tool', {localeDir: 1})).
      toThrow(
        'Invalid parameter. Expected localeDir to be a string but found number instead');
  });

  test('missing locale dir', () => {
    global.console = { warn: jest.fn(), info: jest.fn() };
    const ConnectedComponent = connectTool('tool', {})(TestComponent);
    const wrapper = mount(
      <ConnectedComponent.container writeProjectData={jest.fn()}
                                    readProjectData={jest.fn()}
                                    readProjectDataSync={jest.fn()}
                                    deleteProjectFile={jest.fn()}
                                    projectDataPathExists={jest.fn()}
                                    projectDataPathExistsSync={jest.fn()}
                                    appLanguage="en_US"/>
    );
    const component = wrapper.find('TestComponent');
    expect(component.props().tc.appLanguage).toEqual('en_US');
    expect(component.props().currentLanguage).not.toBeDefined();
    expect(component.props().translate).not.toBeDefined();
    expect(console.warn).toBeCalled();
  });

  test('with locale', () => {
    global.console = { info: jest.fn() };
    const ConnectedComponent = connectTool('tool', {localeDir})(TestComponent);
    const wrapper = mount(
      <ConnectedComponent.container writeProjectData={jest.fn()}
                                    readProjectData={jest.fn()}
                                    readProjectDataSync={jest.fn()}
                                    deleteProjectFile={jest.fn()}
                                    projectDataPathExists={jest.fn()}
                                    projectDataPathExistsSync={jest.fn()}
                                    appLanguage="de_DE"/>
    );
    const component = wrapper.find('TestComponent');
    expect(component.props().tc.appLanguage).toEqual('de_DE');
    expect(component.props().currentLanguage).toEqual('de_DE');
    expect(component.props().translate).toBeDefined();
  });

  test('missing app language', () => {
    global.console = { error: jest.fn(), info: jest.fn() };
    const ConnectedComponent = connectTool('tool', {localeDir})(TestComponent);
    const wrapper = mount(
      <ConnectedComponent.container writeProjectData={jest.fn()}
                                    readProjectData={jest.fn()}
                                    readProjectDataSync={jest.fn()}
                                    deleteProjectFile={jest.fn()}
                                    projectDataPathExists={jest.fn()}
                                    projectDataPathExistsSync={jest.fn()}/>
    );
    const component = wrapper.find('TestComponent');
    expect(component.props().appLanguage).not.toBeDefined();
    expect(component.props().currentLanguage).toEqual('en_US');
    expect(component.props().translate).toBeDefined();
    expect(console.error).toBeCalled();
  });
});

describe('snapshots', () => {

  function onError(e) {
    e.preventDefault();
  }

  beforeEach(() => {
    window.addEventListener('error', onError);
  });

  afterEach(() => {
    window.removeEventListener('error', onError);
  });

  it('renders the wrapped component', () => {
    const ConnectedComponent = connectTool('tool', {localeDir})(TestComponent);
    const wrapper = renderer.create(
      <ConnectedComponent.container writeProjectData={jest.fn()}
                                    readProjectData={jest.fn()}
                                    readProjectDataSync={jest.fn()}
                                    deleteProjectFile={jest.fn()}
                                    projectDataPathExists={jest.fn()}
                                    projectDataPathExistsSync={jest.fn()}
                                    appLanguage="de_DE"/>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
