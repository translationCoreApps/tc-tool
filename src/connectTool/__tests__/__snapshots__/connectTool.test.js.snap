// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots renders the wrapped component 1`] = `
<div>
  Hello World!
</div>
`;
